Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/provider.go

```diff
@@ -41,45 +51,81 @@ func NewSimpleProvider(scc *api.SecurityContextConstraints) (SecurityContextCons
 		return nil, fmt.Errorf("NewSimpleProvider requires a SecurityContextConstraints")
 	}
 
-	var userStrat user.RunAsUserSecurityContextConstraintsStrategy = nil
-	var err error = nil
-	switch scc.RunAsUser.Type {
-	case api.RunAsUserStrategyMustRunAs:
-		userStrat, err = user.NewMustRunAs(&scc.RunAsUser)
-	case api.RunAsUserStrategyMustRunAsRange:
-		userStrat, err = user.NewMustRunAsRange(&scc.RunAsUser)
-	case api.RunAsUserStrategyMustRunAsNonRoot:
-		userStrat, err = user.NewRunAsNonRoot(&scc.RunAsUser)
-	case api.RunAsUserStrategyRunAsAny:
-		userStrat, err = user.NewRunAsAny(&scc.RunAsUser)
-	default:
-		err = fmt.Errorf("Unrecognized RunAsUser strategy type %s", scc.RunAsUser.Type)
+	userStrat, err := createUserStrategy(&scc.RunAsUser)
+	if err != nil {
+		return nil, err
 	}
+
+	seLinuxStrat, err := createSELinuxStrategy(&scc.SELinuxContext)
 	if err != nil {
 		return nil, err
 	}
 
-	var seLinuxStrat selinux.SELinuxSecurityContextConstraintsStrategy = nil
-	err = nil
-	switch scc.SELinuxContext.Type {
-	case api.SELinuxStrategyMustRunAs:
-		seLinuxStrat, err = selinux.NewMustRunAs(&scc.SELinuxContext)
-	case api.SELinuxStrategyRunAsAny:
-		seLinuxStrat, err = selinux.NewRunAsAny(&scc.SELinuxContext)
-	default:
-		err = fmt.Errorf("Unrecognized SELinuxContext strategy type %s", scc.SELinuxContext.Type)
+	fsGroupStrat, err := createFSGroupStrategy(&scc.FSGroup)
+	if err != nil {
+		return nil, err
 	}
+
+	supGroupStrat, err := createSupplementalGroupStrategy(&scc.SupplementalGroups)
 	if err != nil {
 		return nil, err
 	}
 
 	return &simpleProvider{
-		scc:               scc,
-		runAsUserStrategy: userStrat,
-		seLinuxStrategy:   seLinuxStrat,
+		scc:                       scc,
+		runAsUserStrategy:         userStrat,
+		seLinuxStrategy:           seLinuxStrat,
+		fsGroupStrategy:           fsGroupStrat,
+		supplementalGroupStrategy: supGroupStrat,
 	}, nil
 }
 
+// Create a PodSecurityContext based on the given constraints.  If a settingis already set
+// on the PodSecurityContext it will not be changed.  Validate should be used after the context
+// is created to ensureit complies with the required restrictions.
+//
+// NOTE: this method works on a copy of the PodSecurityContext.  It is up to the caller to
+// apply the PSC if validation passes.
+func (s *simpleProvider) CreatePodSecurityContext(pod *api.Pod) (*api.PodSecurityContext, error) {
+	var sc *api.PodSecurityContext = nil
+	if pod.Spec.SecurityContext != nil {
+		// work with a copy
+		copy := *pod.Spec.SecurityContext
+		sc = &copy
+	} else {
+		sc = &api.PodSecurityContext{}
+	}
+
+	if len(sc.SupplementalGroups) == 0 {
+		supGroups, err := s.supplementalGroupStrategy.Generate(pod)
+		if err != nil {
+			return nil, err
+		}
+		sc.SupplementalGroups = supGroups
+	}
+
+	if sc.FSGroup == nil {
+		fsGroup, err := s.fsGroupStrategy.Generate(pod)
```

Comment:
I want to be able to reuse the same scc strategies for sup group and fsgroup.  This allows custom strategies to provide ranges of acceptable groups and multiple defaults.  I thought it best to leave the interface flexible enough to provide multiple values.

Two things to note:

1.  fsgroup is a single field (so to reuse I need to make sure I only take a single value)
2.  the default mustRunAs generation is just returning the first value in the configured ranges anyway (but as mentioned, I thought it best to leave the interface capable of returning multiple to support defaulting to multiple sup groups)